compiler/src/dmd/expressionsem.d

```diff
@@ -1301,7 +1301,19 @@ private Expression resolvePropertiesX(Scope* sc, Expression e1, Expression e2 =
                     }
                     Expression e = new CallExp(loc, e1);
                     if (e2)
-                        e = new AssignExp(loc, e, e2);
+                    {
+                        /* https://issues.dlang.org/show_bug.cgi?id=23335
+                         *
+                         * If we have an assignement between FuncExps,
+                         * the type of e2 needs to be deduced from e1.
+                         * However, by this point e2 has already been
+                         * semantically evaluated.
+                         * To make sure that AssignExp semantic goes well
+                         * start from a clean slate with e2 by performing
+                         * a syntax copy.
+                         */
+                        e = new AssignExp(loc, e, e2.isFuncExp() ? e2.syntaxCopy() : e2);
```

Comment:
What happens to the original version of `e2`? What if there are already references to it elsewhere in the AST?

Reply:
That cannot happen. The only situations were `resolvePropertiesX` is called with a non-null `e2` are [1] and [2]. Both cases are in `AssignExp` semantic which is the only entity that references `e2`; in both situations, the result of `resolvePropertiesX` is returned, therefore, the syntaxCopy is entirely safe.

[1] https://github.com/dlang/dmd/blob/master/compiler/src/dmd/expressionsem.d#L8864
[2] https://github.com/dlang/dmd/blob/master/compiler/src/dmd/expressionsem.d#L8889

Reply:
@ibuclaw Is this ok?